File: OrchidCore/src/main/java/com/eden/orchid/api/server/OrchidServer.java
package com.eden.orchid.api.server;

import com.eden.orchid.api.OrchidContext;
import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.IOException;
import java.util.Set;

@Singleton
public class OrchidServer {
    private final OrchidContext context;
    private int httpServerPort;
    private int websocketPort;
    public OrchidWebserver server;
    public OrchidWebsocket websocket;
    private final Set<OrchidController> controllers;
    private final OrchidFileController fileController;

    @Inject
    public OrchidServer(OrchidContext context, Set<OrchidController> controllers, OrchidFileController fileController) {
        this.context = context;
        this.controllers = controllers;
        this.fileController = fileController;
    }

    public void start(int port) throws IOException {
        // TODO: setup assistedInject for these
        this.server = new OrchidWebserver(context, controllers, fileController, port);
        this.httpServerPort = this.server.getListeningPort();
        this.websocket = new OrchidWebsocket(context, port);
        this.websocketPort = this.websocket.getListeningPort();
    }

    public OrchidContext getContext() {
        return this.context;
    }

    public int getHttpServerPort() {
        return this.httpServerPort;
    }

    public int getWebsocketPort() {
        return this.websocketPort;
    }

    public OrchidWebserver getServer() {
        return this.server;
    }

    public OrchidWebsocket getWebsocket() {
        return this.websocket;
    }

    public Set<OrchidController> getControllers() {
        return this.controllers;
    }

    public OrchidFileController getFileController() {
        return this.fileController;
    }
}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/server/OrchidServer.kt
package com.eden.orchid.api.server

import com.eden.orchid.api.OrchidContext
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
@JvmSuppressWildcards
class OrchidServer
@Inject
constructor(
    val controllers: Set<OrchidController>,
    val fileController: OrchidFileController
) {
    var httpServerPort = 0
    var websocketPort = 0

    var server: OrchidWebserver? = null
    var websocket: OrchidWebsocket? = null

    @Throws(IOException::class)
    fun start(context: OrchidContext, port: Int) {
        server = OrchidWebserver(context, controllers, fileController, port)
        httpServerPort = server!!.listeningPort

        websocket = OrchidWebsocket(context, port)
        websocketPort = websocket!!.listeningPort
    }

}
